framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C

```diff
@@ -54,6 +54,26 @@ DistributedRectilinearMeshGenerator::validParams()
   params.addParam<processor_id_type>(
       "num_cores_for_partition", 0, "Number of cores for partitioning the graph");
 
+  params.addParam<unsigned>(
+      "num_side_layers",
+      2,
+      "Number of layers of off-processor side neighbors is reserved during mesh generation");
+
+  params.addRelationshipManager(
+      "ElementSideNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC,
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
+        // Let this RM safeguard users specified ghosted layers
+        rm_params.set<unsigned short>("layers") = obj_params.get<unsigned>("num_side_layers");
+        // We can not attach geometric early here because some simulation related info, such as,
+        // periodic BCs is not available yet during an early stage. Periodic BCs will be passed
+        // into ghosting functor during initFunctor of ElementSideNeighborLayers. There is no hurt
+        // to attach geometric late for general simulations that do not have extra requirements on
+        // ghosting elements. That is especially true distributed generated meshes since there is
+        // not much redundant info.
+        rm_params.set<bool>("attach_geometric_early") = false;
+      });
```

Comment:
I'm curious whether this is necessary. When I think about a ghosting need, I (almost) always think there's an algebraic or coupling driver. For instance I wouldn't request that my mesh keep 10 layers of neighbors ghosted just for fun; I would only do so if I plan on accessing vector or matrix entries that are 10 layers out. It seems that the motivator for this PR is PolycrystalVoronoi. If I look in `PolycrystalUserObjectBase`, I see that we should already be adding two layers of neighbor side geometric ghosting. I'm guessing that you still probably need your `num_side_layers` parameter here, but I'm just thinking you can probably leave off addition of the geometric RM here. I might even suggest making `num_side_layers` a required parameter such that users have to think about what they are doing, e.g. what objects do they have in their simulation that require off-process side neighbors.

Reply:
`PolycrystalVoronoi` or `grain tracker` requires two layers of geometric ghosting elements (by default), but we need only one layer of coupling and algebraic elements.   That being said, there are use cases where a geometric RM can not be overridden by other coupling and algebraic RMs. 

I do like keeping `the addition of the geometric RM` in the mesh generator.   If there is no addition of a geometric RM here, we can not implement `num_side_layers` correctly. For example, if I have `num_side_layers=3`, then go to run the diffusion.i, and then we will end up with results of `num_side_layers=1`.

Reply:
Ok, I am fine with keeping the geometric RM addition here. I'm curious: is there a problem if a user specifies `num_side_layers=1` in the mesh generator but then they run `PolyCrystalVoronoi` or the grain tracker? If I understand correctly `num_sides_layers` controls how many side layers of elements you *generate* so if you don't generate sufficient side layers, then you will be in trouble?

If my above scenario is indeed a problem, then I don't have any bright idea for fixing it automatically for the user. At that point maybe we just say it is the user's fault.

Reply:
Yes, good point. `num_sides_layers` is used to control we have sufficient ghosting elements at the first place. 

That is an egg-chicken loop here. Without a mesh, we can not create kernels, UOs, etc. Without knowing RMs in kernels, MOs, etc. we have no idea how much information we need to generate in the first place.

Looking forward someday we will have an intelligent  RM that can grow itself automatically.  If such an intelligent  RM exists, then we do not even to generate overlapping domain at the first place